### PR TITLE
Dictionary object model

### PR DESCRIPTION
**Pull request by syndacks**
Added a "DictionaryAsset" to index.js. This will support the "DictionaryObject" data model that is for DiscoveryFeed. I also added some related tests in verify.js.

https://phabricator.endlessm.com/T17329

### CODE REVIEW

**Comment by sgnn7** [comment]
@syndacks Don't forget about https://github.com/endlessm/libingester/blob/master/previewer/metadata_renderer.js#L7

Also, squash your commits.

**Comment by syndacks** [comment]
@sgnn7 Thanks, reviewed and squashed

File: lib/index.js [hunk 1]
```diff
@@ -178,6 +178,35 @@ function _ensure_cheerio(value) {
...
 
+class DictionaryAsset extends BaseAsset {
+    constructor() {
+        super();
+        this._object_type = "DictionaryObject";
+        this._content_type = "text/html";
+    }
+
+    set_word (value) { this._word = value; }
+    set_wordDefinition (value) { this._wordDefinition = value; }
```
Comment (ptomato) [inline lib/index.js]: drive-by nitpick: don't mix `camelCase` and `snake_case`

Comment (syndacks) [inline lib/index.js]: @ptomato thanks! please continue to drive/walk/run-by any code of mine and offer review.

Comment (sgnn7) [inline lib/index.js]: +1 on casing

File: lib/index.js [hunk 2]
```diff
@@ -178,6 +178,35 @@ function _ensure_cheerio(value) {
...
+    set_partOfSpeech (value) { this._partOfSpeech = value; }
+    set_tags (value) { this._tags = value; }
+
+    to_metadata() {
+        const metadata = super.to_metadata();
+        Object.assign(metadata, {
+            'word': this._word,
+            'wordDefinition': this._wordDefinition,
+            'partOfSpeech': this._partOfSpeech,
```
Comment (sgnn7) [inline lib/index.js]: Ditto on casing

File: lib/index.js [hunk 3]
```diff
@@ -178,6 +178,35 @@ function _ensure_cheerio(value) {
...
+            'tags': [this._tags]
+        });
+        console.log(metadata);
```
Comment (sgnn7) [inline lib/index.js]: Probably needs to be removed

File: lib/verify.js [hunk 4]
```diff
@@ -30,6 +30,8 @@ function verify_metadata(metadata) {
         assert(!!metadata['document'], "metadata missing document");
     } else if (object_type === 'ImageObject') {
         assert(!!metadata['cdnFilename'], "Image object missing cdnFilename (has no data?)");
+    } else if (object_type === "DictionaryObject") {
+        assert(!!metadata['word'], "metadata missing word");
```
Comment (sgnn7) [inline lib/verify.js]: Not checking that word definition and part of speech is set as well?

File: previewer/metadata_renderer.js [hunk 5]
```diff
@@ -7,7 +7,8 @@ hatchFolder = main.hatchFolder
 const extraPropsByModelType = {
   'ArticleObject': ['authors', 'published', 'sourceName'],
   'ImageObject': ['caption', 'height', 'width'],
-  'VideoObject': ['caption', 'duration', 'height', 'transcript', 'width']
+  'VideoObject': ['caption', 'duration', 'height', 'transcript', 'width'],
+  'DictionaryObject': ['word', 'wordDefinition', 'partOfSpeech']
```
Comment (sgnn7) [inline previewer/metadata_renderer.js]: Ditto on casing

File: lib/verify.js (hunk 4 above)
Comment (syndacks) [inline lib/verify.js]: Good point, I'm going to leave off "part of speech" as that doesn't seem as important as a word/wordDefinition pair.

File: lib/index.js [hunk 6]
```diff
@@ -178,6 +178,34 @@ function _ensure_cheerio(value) {
...
+    set_part_of_speech (value) { this._part_of_speech = value; }
+    set_tags (value) { this._tags = value; }
+
+    to_metadata() {
+        const metadata = super.to_metadata();
+        Object.assign(metadata, {
+            'word': this._word,
+            'wordDefinition': this._word_definition,
+            'partOfSpeech': this._part_of_speech,
+            'tags': [this._tags]
```
Comment (wgreenberg) [inline lib/index.js]: this line shouldn't be necessary, also `_tags` as a variable implies it should be an array

File: lib/index.js [hunk 7]
```diff
@@ -178,6 +178,34 @@ function _ensure_cheerio(value) {
     return value;
 }
 
+class DictionaryAsset extends BaseAsset {
+    constructor() {
+        super();
+        this._object_type = "DictionaryObject";
+        this._content_type = "text/html";
```
Comment (wgreenberg) [inline lib/index.js]: I don't think this is actually correct, since afaik there's no HTML involved here. To be honest I'm not sure what the content type of a "word" is

Comment (ptomato) [inline lib/index.js]: I might suggest having some sort of rendered HTML in the document body anyway, even if it's not used directly in the discovery feed...

File: lib/index.js (hunk 6 above)
Comment (syndacks) [inline lib/index.js]: I had to explicitly set the tags as an array to match our `verify` condition.

File: lib/index.js (hunk 7 above)
Comment (syndacks) [inline lib/index.js]: Sounds good - there's a `render` method now in the DictionaryAsset class which sets the relevant fields (word, wordDefinition, partOfSpeech) to already existing fields on the Blog template.

Comment (syndacks) [inline lib/index.js]: @wgreenberg let me know what you think

Comment (ptomato) [inline lib/index.js]: I would not reuse the blog template for this - just make some simple HTML document with no styling for the time being

Comment (syndacks) [inline lib/index.js]: thanks @ptomato - changed made.

File: lib/index.js [hunk 8]
```diff
@@ -183,6 +183,48 @@ function _ensure_cheerio(value) {
...
+    set_definition (value) { this._definition = value; }
+    set_part_of_speech (value) { this._part_of_speech = value; }
+    set_tags (value) { this._tags = value; }
+
+    to_metadata() {
+        const metadata = super.to_metadata();
+        Object.assign(metadata, {
+            'word': this._word,
+            'definition': this._definition,
+            'part-of-speech': this._part_of_speech,
```
Comment (wgreenberg) [inline lib/index.js]: in the other classes, we use camelCase for metadata keys. s/part-of-speech/partOfSpeech

File: previewer/metadata_renderer.js [hunk 9]
```diff
@@ -7,7 +7,8 @@ hatchFolder = main.hatchFolder
 const extraPropsByModelType = {
   'ArticleObject': ['authors', 'published', 'sourceName'],
   'ImageObject': ['caption', 'height', 'width'],
-  'VideoObject': ['caption', 'duration', 'height', 'transcript', 'width']
+  'VideoObject': ['caption', 'duration', 'height', 'transcript', 'width'],
+  'DictionaryObjectModel': ['word', 'definition', 'part-of-speech']
```
Comment (wgreenberg) [inline previewer/metadata_renderer.js]: nit: trailing comma

Comment (wgreenberg) [inline previewer/metadata_renderer.js]: also change to partOfSpeech

File: lib/index.js (hunk 8 above)
Comment (syndacks) [inline lib/index.js]: General question...these metadata fields need to match up with: https://github.com/endlessm/endless-content-factory/pull/742

right?

**Comment by cosimoc** [comment]
Needs rebase to master.

**Comment by syndacks** [comment]
@cosimoc rebased
@wgreenberg ready for review

File: lib/index.js [hunk 10]
```diff
@@ -204,6 +204,48 @@ class VideoAsset extends BaseAsset {
...
+    set_word (value) { this._word = value; }
+    set_definition (value) { this._definition = value; }
+    set_part_of_speech (value) { this._part_of_speech = value; }
+    set_tags (value) { this._tags = value; }
+
+    to_metadata() {
+        const metadata = super.to_metadata();
+        Object.assign(metadata, {
+            'word': this._word,
+            'definition': this._definition,
```
Comment (wgreenberg) [inline lib/index.js]: shouldn't this be wordDefinition?

Comment (syndacks) [inline lib/index.js]: @wgreenberg I think @smspillaz changed this?...i'm just about to head out for the night so can take a look on Monday

Comment (smspillaz) [inline lib/index.js]: This should be `definition` - see https://github.com/endlessm/endless-content-factory/pull/742/commits/5dc7dcdef80e9e93aecfe5f6fcb7c7d63db4336d

Unless there's something else that needs to change?

Comment (syndacks) [inline lib/index.js]: `definition` confirmed correct

**Comment by wgreenberg** [comment]
LGTM
